pallets/native-token-management/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+//!            it is not a fully-featured token bridge and needs to be combined with a separate
+//!            sender-receiver metadata channel to implement one.
+//!
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::inherent::IsFatalError;
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sidechain_domain::*;
+use sp_native_token_management::*;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(test, feature = "mock"))]
+mod mock;
+
+#[derive(Encode, Debug, PartialEq)]
+pub enum InherentError {
+	TokenTransferNotHandled,
+	IncorrectTokenNumberTransfered,
+	UnexpectedTokenTransferInherent,
+}
+
+impl IsFatalError for InherentError {
+	fn is_fatal_error(&self) -> bool {
+		true
+	}
+}
+
+/// Interface for user-provided logic to handle native token transfers into the illiquid supply on the main chain.
+///
+/// The handler will be called with **the total sum** of transfers since the previous partner chain block.
+pub trait TokenTransferHandler {
+	fn handle_token_transfer(token_amount: NativeTokenAmount) -> DispatchResult;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		type TokenTransferHandler: TokenTransferHandler;
+	}
+
+	#[pallet::event]
+	pub enum Event<T: Config> {
+		TokensTransfered(NativeTokenAmount),
+	}
+
+	#[pallet::storage]
+	pub type MainChainScriptsConfiguration<T: Config> =
+		StorageValue<_, sp_native_token_management::MainChainScripts, ValueQuery>;
+
+	#[pallet::genesis_config]
+	#[derive(frame_support::DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		pub main_chain_scripts: sp_native_token_management::MainChainScripts,
+		pub _marker: PhantomData<T>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			MainChainScriptsConfiguration::<T>::put(self.main_chain_scripts.clone());
+		}
+	}
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+			Self::get_transfered_tokens_from_inherent_data(data)
+				.filter(|data| !data.token_amount.0 > 0)
+				.map(|data| Call::transfer_tokens { token_amount: data.token_amount })
+		}
+
+		fn check_inherent(call: &Self::Call, data: &InherentData) -> Result<(), Self::Error> {
+			let Some(token_transfer) = Self::get_transfered_tokens_from_inherent_data(data) else {
+				return Err(InherentError::UnexpectedTokenTransferInherent);
+			};
+			let Call::transfer_tokens { token_amount } = call else {
+				unreachable!("There is no other extrinsic in this pallet");
```

Comment:
`set_main_chain_scripts` is the other extrinsic in this pallet.

Reply:
Right, I added it later. Using `match` would have made the compiler catch that change. changed that to return `Ok(())` like other `check_inherent`s